File: benchmarks/BenchStress.cs
// ============================================================
// C# PERFORMANCE BENCHMARK SUITE
// All times in milliseconds
// ============================================================

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;

class BenchStress
{
    static long Fib(int n)
    {
        long a = 0, b = 1;
        for (int i = 0; i < n; i++)
        {
            long tmp = a + b;
            a = b;
            b = tmp;
        }
        return a;
    }

    static long MatmulSim(int size)
    {
        long sum = 0;
        for (int i = 0; i < size; i++)
            for (int j = 0; j < size; j++)
                for (int k = 0; k < size; k++)
                    sum = sum + (i * k) - (j * k) + 1;
        return sum;
    }

    static int CountPrimes(int limit)
    {
        int count = 0;
        for (int n = 2; n < limit; n++)
        {
            bool isPrime = true;
            for (int d = 2; d * d <= n; d++)
            {
                if (n % d == 0) { isPrime = false; break; }
            }
            if (isPrime) count++;
        }
        return count;
    }

    static long ArrayStress(int n)
    {
        var arr = new List<long>();
        for (int i = 0; i < n; i++)
            arr.Add((long)i * i);
        long sum = 0;
        for (int i = 0; i < arr.Count; i++)
            sum += arr[i];
        return sum;
    }

    static long MapStress(int n)
    {
        var m = new Dictionary<string, long>();
        for (int i = 0; i < n; i++)
            m["key"] = (long)i * i;
        long sum = 0;
        foreach (var key in m.Keys)
            sum += m[key];
        return sum;
    }

    class Vector
    {
        public long X, Y;
        public Vector(long x, long y) { X = x; Y = y; }
        public Vector Add(Vector other) => new Vector(X + other.X, Y + other.Y);
        public long MagnitudeSquared() => X * X + Y * Y;
    }

    static long OopStress(int n)
    {
        long total = 0;
        for (int i = 0; i < n; i++)
        {
            var v1 = new Vector(i, i + 1);
            var v2 = new Vector(i + 2, i + 3);
            var v3 = v1.Add(v2);
            total += v3.MagnitudeSquared();
        }
        return total;
    }

    class Shape
    {
        public string Name;
        public Shape(string name) { Name = name; }
        public virtual long Area() => 0;
    }

    class Rectangle : Shape
    {
        public long W, H;
        public Rectangle(long w, long h) : base("rect") { W = w; H = h; }
        public override long Area() => W * H;
    }

    static long InheritanceStress(int n)
    {
        long total = 0;
        for (int i = 0; i < n; i++)
        {
            var r = new Rectangle(i, i + 1);
            total += r.Area();
        }
        return total;
    }

    static Func<int> MakeCounter()
    {
        int count = 0;
        return () => { count++; return count; };
    }

    static int ClosureStress(int n)
    {
        var counter = MakeCounter();
        int result = 0;
        for (int i = 0; i < n; i++)
            result = counter();
        return result;
    }

    static int StringStress(int n)
    {
        var sb = new StringBuilder();
        for (int i = 0; i < n; i++)
            sb.Append("x");
        return sb.Length;
    }

    static int FibRecursive(int n)
    {
        if (n <= 1) return n;
        return FibRecursive(n - 1) + FibRecursive(n - 2);
    }

    static long Ms(Stopwatch sw) => sw.ElapsedMilliseconds;

    static void Main()
    {
        Console.WriteLine("=== C# PERFORMANCE BENCHMARK SUITE ===");

        var sw = Stopwatch.StartNew();
        var t0 = sw.ElapsedMilliseconds;

        Fib(1000000);
        var t1 = sw.ElapsedMilliseconds;
        Console.WriteLine("1_fib_iter_1M");
        Console.WriteLine(t1 - t0);

        var t2 = sw.ElapsedMilliseconds;
        MatmulSim(100);
        var t3 = sw.ElapsedMilliseconds;
        Console.WriteLine("2_nested_loops_1M");
        Console.WriteLine(t3 - t2);

        var t4 = sw.ElapsedMilliseconds;
        CountPrimes(50000);
        var t5 = sw.ElapsedMilliseconds;
        Console.WriteLine("3_primes_50K");
        Console.WriteLine(t5 - t4);

        var t6 = sw.ElapsedMilliseconds;
        ArrayStress(100000);
        var t7 = sw.ElapsedMilliseconds;
        Console.WriteLine("4_array_100K");
        Console.WriteLine(t7 - t6);

        var t8 = sw.ElapsedMilliseconds;
        MapStress(50000);
        var t9 = sw.ElapsedMilliseconds;
        Console.WriteLine("5_map_50K");
        Console.WriteLine(t9 - t8);

        var t10 = sw.ElapsedMilliseconds;
        OopStress(100000);
        var t11 = sw.ElapsedMilliseconds;
        Console.WriteLine("6_oop_100K");
        Console.WriteLine(t11 - t10);

        var t12 = sw.ElapsedMilliseconds;
        InheritanceStress(100000);
        var t13 = sw.ElapsedMilliseconds;
        Console.WriteLine("7_inherit_100K");
        Console.WriteLine(t13 - t12);

        var t14 = sw.ElapsedMilliseconds;
        ClosureStress(1000000);
        var t15 = sw.ElapsedMilliseconds;
        Console.WriteLine("8_closure_1M");
        Console.WriteLine(t15 - t14);

        var t16 = sw.ElapsedMilliseconds;
        StringStress(10000);
        var t17 = sw.ElapsedMilliseconds;
        Console.WriteLine("9_string_10K");
        Console.WriteLine(t17 - t16);

        var t18 = sw.ElapsedMilliseconds;
        FibRecursive(30);
        var t19 = sw.ElapsedMilliseconds;
        Console.WriteLine("10_fib_rec_30");
        Console.WriteLine(t19 - t18);

        Console.WriteLine("TOTAL");
        Console.WriteLine(t19 - t0);
    }
}

File: benchmarks/bench_stress.go
package main

import (
	"fmt"
	"strings"
	"time"
)

func fib(n int) int {
	a, b := 0, 1
	for i := 0; i < n; i++ {
		a, b = b, a+b
	}
	return a
}

func matmulSim(size int) int {
	sum := 0
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			for k := 0; k < size; k++ {
				sum = sum + (i*k) - (j*k) + 1
			}
		}
	}
	return sum
}

func countPrimes(limit int) int {
	count := 0
	for n := 2; n < limit; n++ {
		isPrime := true
		for d := 2; d*d <= n; d++ {
			if n%d == 0 {
				isPrime = false
				break
			}
		}
		if isPrime {
			count++
		}
	}
	return count
}

func arrayStress(n int) int {
	arr := make([]int, 0, n)
	for i := 0; i < n; i++ {
		arr = append(arr, i*i)
	}
	sum := 0
	for _, v := range arr {
		sum += v
	}
	return sum
}

func mapStress(n int) int {
	m := make(map[string]int)
	for i := 0; i < n; i++ {
		m["key"] = i * i
	}
	sum := 0
	for _, v := range m {
		sum += v
	}
	return sum
}

type Vector struct {
	x, y int
}

func NewVector(x, y int) *Vector { return &Vector{x, y} }

func (v *Vector) Add(other *Vector) *Vector {
	return NewVector(v.x+other.x, v.y+other.y)
}

func (v *Vector) MagnitudeSquared() int {
	return v.x*v.x + v.y*v.y
}

func oopStress(n int) int {
	total := 0
	for i := 0; i < n; i++ {
		v1 := NewVector(i, i+1)
		v2 := NewVector(i+2, i+3)
		v3 := v1.Add(v2)
		total += v3.MagnitudeSquared()
	}
	return total
}

type Shape struct {
	name string
}

func NewShape(name string) *Shape { return &Shape{name} }
func (s *Shape) Area() int       { return 0 }

type Rectangle struct {
	*Shape
	w, h int
}

func NewRectangle(w, h int) *Rectangle {
	return &Rectangle{Shape: NewShape("rect"), w: w, h: h}
}

func (r *Rectangle) Area() int {
	return r.w * r.h
}

func inheritanceStress(n int) int {
	total := 0
	for i := 0; i < n; i++ {
		r := NewRectangle(i, i+1)
		total += r.Area()
	}
	return total
}

func makeCounter() func() int {
	count := 0
	return func() int {
		count++
		return count
	}
}

func closureStress(n int) int {
	counter := makeCounter()
	result := 0
	for i := 0; i < n; i++ {
		result = counter()
	}
	return result
}

func stringStress(n int) int {
	var sb strings.Builder
	for i := 0; i < n; i++ {
		sb.WriteString("x")
	}
	return sb.Len()
}

func fibRecursive(n int) int {
	if n <= 1 {
		return n
	}
	return fibRecursive(n-1) + fibRecursive(n-2)
}

func ms(d time.Duration) int64 { return d.Milliseconds() }

func main() {
	fmt.Println("=== GO PERFORMANCE BENCHMARK SUITE ===")

	t0 := time.Now()
	fib(1000000)
	t1 := time.Now()
	fmt.Println("1_fib_iter_1M")
	fmt.Println(ms(t1.Sub(t0)))

	t2 := time.Now()
	matmulSim(100)
	t3 := time.Now()
	fmt.Println("2_nested_loops_1M")
	fmt.Println(ms(t3.Sub(t2)))

	t4 := time.Now()
	countPrimes(50000)
	t5 := time.Now()
	fmt.Println("3_primes_50K")
	fmt.Println(ms(t5.Sub(t4)))

	t6 := time.Now()
	arrayStress(100000)
	t7 := time.Now()
	fmt.Println("4_array_100K")
	fmt.Println(ms(t7.Sub(t6)))

	t8 := time.Now()
	mapStress(50000)
	t9 := time.Now()
	fmt.Println("5_map_50K")
	fmt.Println(ms(t9.Sub(t8)))

	t10 := time.Now()
	oopStress(100000)
	t11 := time.Now()
	fmt.Println("6_oop_100K")
	fmt.Println(ms(t11.Sub(t10)))

	t12 := time.Now()
	inheritanceStress(100000)
	t13 := time.Now()
	fmt.Println("7_inherit_100K")
	fmt.Println(ms(t13.Sub(t12)))

	t14 := time.Now()
	closureStress(1000000)
	t15 := time.Now()
	fmt.Println("8_closure_1M")
	fmt.Println(ms(t15.Sub(t14)))

	t16 := time.Now()
	stringStress(10000)
	t17 := time.Now()
	fmt.Println("9_string_10K")
	fmt.Println(ms(t17.Sub(t16)))

	t18 := time.Now()
	fibRecursive(30)
	t19 := time.Now()
	fmt.Println("10_fib_rec_30")
	fmt.Println(ms(t19.Sub(t18)))

	fmt.Println("TOTAL")
	fmt.Println(ms(t19.Sub(t0)))
}
